cmd/maya-exporter/app/collector/cstor.go

```diff
@@ -0,0 +1,211 @@
+	// EOF separates the strings from response which comes from the
+	// cstor as the collection of metrics.
+	EOF = "\r\n"
+	// Footer is used to verify if all the response has collected.
+	Footer = "OK IOSTATS"
+	// Command is a command that is used to write over wire and get
+	// the iostats from the cstor.
+	Command = "IOSTATS"
+	// BufSize is the size of response from cstor.
+	BufSize = 1024
+)
+
+// cstor implements the Exporter interface. It exposes
+// the metrics of a OpenEBS (cstor) volume.
+type cstor struct {
+	conn  net.Conn
+	stats stats
+}
+
+func Cstor() *cstor {
+	return &cstor{
+		stats: stats{},
+	}
+}
+
+// InitiateConnection tries to initiates the connection with the cstor
+// over unix domain socket. This function can not be unit tested (only
+// negative cases are possible).
+func (c *cstor) InitiateConnection(path string) error {
+	conn, err := net.Dial("unix", path)
+	if err != nil {
+		return err
+	}
+	if conn != nil {
+		c.conn = conn
+		glog.Info("Connection established with istgt")
+		c.readHeader()
+	}
+	return nil
+}
+
+// writer writes the "IOSTATS\n" command on the socket. it returns
+// error if the connection is closed or not available with socket.
+func (c *cstor) writer() error {
+	msg := Command + "\n"
+	_, err := c.conn.Write([]byte(msg))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// reader reads the response from socket in the buffer and
+// exits if the buffer contains the "IOSTATS\r\n" at the end.
+func (c *cstor) reader() (string, error) {
+	buf := make([]byte, BufSize)
+	var (
+		err           error
+		n             int
+		str, response string
+		buffer        bytes.Buffer
+	)
+	// infinite for loop to collect all the chunks.
+	for {
+		n, err = c.conn.Read(buf[:])
+		if err != nil {
+			return "", err
+		}
+		// concatnat the chunks received and then compare if it has
+		// Footer (" IOSTATS") + EOF ("\r\n") and exit, else continue
+		// appending the chunks at the end of str.
+		buffer.WriteString(string(buf[0:n]))
+		str = buffer.String()
+		if str[len(str)-12:] == Footer+EOF {
+			response = str
+			break
+		}
+	}
+	return response, nil
+}
+
+// ReadHeader only reads the header of the response from cstor
+func (c *cstor) readHeader() error {
+	buf := make([]byte, 1024)
+	var (
+		err    error
+		n      int
+		str    string
+		buffer bytes.Buffer
+	)
+	// collect all the chunks ending with EOF ("\r\n").
+	for {
+		n, err = c.conn.Read(buf[:])
+		if err != nil {
+			glog.Error("Error in reading response, error : ", err)
+			return err
+		}
+		// apend the chunks into str
+		buffer.WriteString(string(buf[0:n]))
+		str = buffer.String()
+		if strings.HasPrefix(str, HeaderPrefix) && strings.HasSuffix(str, EOF) {
+			break
+		}
+	}
+	glog.Infof("Got header: %#v", str)
+	return nil
+}
+
+// getter writes query on the socket for the getting stats
+// if the connection is available else retry to initiate
+// connection again.
+func (c *cstor) getter() (v1.VolumeStats, error) {
```

Comment:
How about `getVolumeStats` as method name?

Reply:
changin to get() only, since getVolumeStats is already a method of Jiva